cf-check/diagnose.c

```diff
@@ -0,0 +1,353 @@
+    macro(LMDB_TRANSACTION_FULL)                          \
+    macro(LMDB_CURSOR_STACK_TOO_DEEP)                     \
+    macro(LMDB_PAGE_FULL)                                 \
+    macro(LMDB_MAP_RESIZE_BEYOND_SIZE)                    \
+    macro(LMDB_INCOMPATIBLE_OPERATION)                    \
+    macro(LMDB_INVALID_REUSE_OF_READER_LOCKTABLE_SLOT)    \
+    macro(LMDB_BAD_OR_INVALID_TRANSACTION)                \
+    macro(LMDB_WRONG_KEY_OR_VALUE_SIZE)                   \
+    macro(LMDB_BAD_DBI)                                   \
+    macro(LMDUMP_UNKNOWN_ERROR)                           \
+    macro(PID_ERROR)                                      \
+    macro(PERMISSION_ERROR)                               \
+    macro(DOES_NOT_EXIST)                                 \
+    macro(UNKNOWN)
+
+#define CF_CHECK_MAX CF_CHECK_UNKNOWN
+
+#define CF_CHECK_CREATE_ENUM(name) \
+  CF_CHECK_##name,
+
+#define CF_CHECK_CREATE_STRING(name) \
+  #name,
+
+typedef enum {
+    CF_CHECK_RUN_CODES(CF_CHECK_CREATE_ENUM)
+} CFCheckCode;
+
+static const char *CF_CHECK_STR[] = {
+    CF_CHECK_RUN_CODES(CF_CHECK_CREATE_STRING)
+};
+
+static bool code_is_errno(int r)
+{
+    return (r > CF_CHECK_MAX);
+}
+
+// Better strerror, returns NULL if it doesn't know.
+static const char *strerror_or_null(int r)
+{
+    const char *strerror_string = strerror(r);
+    if (strerror_string == NULL)
+    {
+        return NULL;
+    }
+
+    const char *unknown = "Unknown error";
+    if (strncmp(strerror_string, unknown, strlen(unknown)) == 0)
+    {
+        return NULL;
+    }
+
+    return strerror_string;
+}
+
+// static bool is_recognized_errno(int r)
+// {
+//     return (strerror_or_null(r) != NULL);
+// }
+
+static int errno_to_code(int r)
+{
+    assert(r != 0);
+    return r + CF_CHECK_MAX;
+}
+
+static int code_to_errno(int r)
+{
+    assert(code_is_errno(r));
+    return r - CF_CHECK_MAX;
+}
+
+static const char *CF_CHECK_STRING(int code)
+{
+    static char unknown[1024];
+    if (code <= 0 || code < CF_CHECK_MAX)
+    {
+        return CF_CHECK_STR[code];
+    }
+    else if (code_is_errno(code)) // code > CF_CHECK_MAX
+    {
+        code = code_to_errno(code);
+        const char *str = strerror_or_null(code);
+        if (str == NULL)
+        {
+            str = "Unknown";
+        }
+        snprintf(unknown, sizeof(unknown), "SYSTEM_ERROR %d - %s", code, str);
+        return unknown;
+    }
+    return CF_CHECK_STR[CF_CHECK_UNKNOWN];
+}
+
+static int signal_to_code(int sig)
+{
+    switch (sig) {
+    case SIGHUP:
+        return CF_CHECK_SIGNAL_HANGUP;
+    case SIGINT:
+        return CF_CHECK_SIGNAL_INTERRUPT;
+    case SIGQUIT:
+        return CF_CHECK_SIGNAL_QUIT;
+    case SIGILL:
+        return CF_CHECK_SIGNAL_ILLEGAL_INSTRUCTION;
+    case SIGTRAP:
+        return CF_CHECK_SIGNAL_TRACE_TRAP;
+    case SIGABRT:
+        return CF_CHECK_SIGNAL_ABORT;
+    // case SIGEMT:
+    //     return CF_CHECK_SIGNAL_EMULATE_INSTRUCTION;
```

Comment:
Please remove this or add a comment about where it appeared.

Reply:
Added a comment